tempesta_fw/vhost.c

```diff
@@ -0,0 +1,608 @@
+ * The directives are deduplicated when put into the array.
+ * Individual directives are linked to from lists of cache policy
+ * directives for specific location sections.
+ */
+#define TFW_CAPOLICY_ARRAY_SZ	(64)
+
+static TfwCaPolicy	tfw_capolicy[TFW_CAPOLICY_ARRAY_SZ];
+static unsigned int	tfw_capolicy_sz = 0;	/* Current size. */
+
+/*
+ * All 'location' directives are put into a fixed size array.
+ * Duplicate directives are not allowed.
+ */
+#define TFW_LOCATION_ARRAY_SZ	(64)
+
+static TfwLocation	tfw_location[TFW_LOCATION_ARRAY_SZ];
+static unsigned int	tfw_location_sz = 0;	/* Current size. */
+
+/*
+ * Default location is a wildcard location. It matches any URI.
+ * It may (or may not) contain a set of cache matching directives.
+ */
+static TfwCaPolicy *tfw_capolicy_dflt[TFW_CAPOLICY_ARRAY_SZ];
+
+static TfwLocation tfw_location_dflt = {
+	.op = TFW_HTTP_MATCH_O_WILDCARD,
+	.arg = "*",
+	.len = 1,
+	.capo = tfw_capolicy_dflt,
+	.capo_sz = 0,
+};
+
+/*
+ * Default vhost is a wildcard vhost. It matches any URI.
+ * It may (or may not) ontain a set of various directives.
+ */
+static TfwVhost		tfw_vhost_dflt = {
+	.loc		= tfw_location,
+	.loc_dflt	= &tfw_location_dflt,
+	.loc_dflt_sz	= 1
+};
+
+/*
+ * Matching functions for match operators. A TfwStr{} is compared
+ * with a plain C string according to a specified match operator.
+ * The functions are generic.
+ */
+static bool
+__tfw_match_wildcard(tfw_match_t op, const char *cstr, int len, TfwStr *arg)
+{
+	return ((op == TFW_HTTP_MATCH_O_WILDCARD)
+		&& (len == 1) && (*cstr == '*'));
+}
+
+static bool
+__tfw_match_suffix(tfw_match_t op, const char *cstr, int len, TfwStr *arg)
+{
+	tfw_str_eq_flags_t flags = TFW_STR_EQ_DEFAULT | TFW_STR_EQ_CASEI;
+	return tfw_str_eq_cstr_off(arg, arg->len - len, cstr, len, flags);
+}
+
+static bool
+__tfw_match_eq(tfw_match_t op, const char *cstr, int len, TfwStr *arg)
+{
+	tfw_str_eq_flags_t flags = TFW_STR_EQ_DEFAULT | TFW_STR_EQ_CASEI;
+	return tfw_str_eq_cstr(arg, cstr, len, flags);
+}
+
+static bool
+__tfw_match_prefix(tfw_match_t op, const char *cstr, int len, TfwStr *arg)
+{
+	tfw_str_eq_flags_t flags = TFW_STR_EQ_PREFIX | TFW_STR_EQ_CASEI;
+	return tfw_str_eq_cstr(arg, cstr, len, flags);
+}
+
+typedef bool (*__tfw_match_fn)(tfw_match_t, const char *, int, TfwStr *);
+
+static const __tfw_match_fn const __read_mostly __tfw_match_fn_tbl[] = {
+	[0 ... _TFW_HTTP_MATCH_O_COUNT] = NULL,
+	[TFW_HTTP_MATCH_O_WILDCARD]	= __tfw_match_wildcard,
+	[TFW_HTTP_MATCH_O_EQ]		= __tfw_match_eq,
+	[TFW_HTTP_MATCH_O_PREFIX]	= __tfw_match_prefix,
+	[TFW_HTTP_MATCH_O_SUFFIX]	= __tfw_match_suffix,
+};
+
+/*
+ * Find a matching cache policy directive. Strings are compared
+ * according to the match operator in the directive. A pointer
+ * to the matching TfwCaPolicy structure is returned if the
+ * match is found. Null is returned if there's no match.
+ */
+static bool
+__tfw_capolicy_match_fn(TfwCaPolicy *capo, TfwStr *arg)
```

Comment:
The function is used only once in `tfw_capolicy_match()`, so there is no sense in the separate function as well as in temporal variables like `match_fn` and `capo`.